### PR TITLE
Dynamic default value

### PR DESCRIPTION
**Pull request by PrettyWood**
## Change Summary
Following #155 and #866, here is a proposal of implementation to simplify dynamic default values

Right now, to have a dynamic default value, we need to use `validators` like
```python
class DynamicValueModel(BaseModel):
    ts: datetime = None

    @validator('ts', pre=True, always=True)
        def set_ts_now(cls, v):
            return v or datetime.now()
```

Now, we can use `default_factory` (same name as [dataclasses](https://docs.python.org/3/library/dataclasses.html#dataclasses.field).to be consistent)
```python
class DynamicValueModel(BaseModel):
    ts: datetime = Field(default_factory=datetime.now)
```

## Related issue number
#155 and #866

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by MrMrRobat** [comment]
Discussed in https://github.com/samuelcolvin/pydantic/issues/866

So I would like to see this:
```py
from datetime import datetime

from pydantic import BaseModel

class Record(BaseModel):
    name: str
    description: str
    ts: datetime = datetime.now

record = Record(name='Foo', description='Bar and Baz also.')
print(record.ts)
#> 2020-02-06 22:11:43.454720
```

However if type of field is callable, it must remain untouched:
```py
class IntGenerator(BaseModel):
    fabric: Callable[[], int] = randint

generator = IntGenerator(name='Foo', description='Bar and Baz also.')
print(generator.fabric(0, 1))
#> 1
```

@samuelcolvin, any thoughts?

**Comment by PrettyWood** [comment]
@MrMrRobat Excellent ! I thought about this implementation but I was afraid it was not wanted. Love it !

File: pydantic/fields.py [hunk 1]
```diff
@@ -76,8 +77,12 @@ class FieldInfo(Representation):
         'extra',
     )
 
-    def __init__(self, default: Any, **kwargs: Any) -> None:
+    def __init__(self, default: Any = Undefined, *, default_factory: Any = None, **kwargs: Any) -> None:
+        if default is not Undefined and default_factory is not None:
```
Comment (samuelcolvin) [inline pydantic/fields.py]: perhaps better to move this check into `Field`m then you can use `pop()` as is done elsewhere.

File: pydantic/fields.py [hunk 2]
```diff
@@ -97,8 +102,9 @@ def __init__(self, default: Any, **kwargs: Any) -> None:
 
 
 def Field(
-    default: Any,
+    default: Any = Undefined,
     *,
+    default_factory: Any = None,
```
Comment (samuelcolvin) [inline pydantic/fields.py]:
```suggestion
    default_factory: Optional[Callable[[], Any]] = None,
```
or something

**Comment by samuelcolvin** [comment]
I'm keen on the approach of

```py
from datetime import datetime

from pydantic import BaseModel

class Record(BaseModel):
    ts: datetime = datetime.now
```

However I think it could cause subtle changes in behaviour in edge cases so let's wait for v2.

Getting `Field(default_factory=...)` to work now is a great first step.

File: tests/test_main.py [hunk 3]
```diff
@@ -970,3 +993,35 @@ class NewModel(DerivedModel, something=2):
...
+
+    m1 = DynamicValueModel()
+    m2 = DynamicValueModel()
+    assert isinstance(m1.uid, UUID)
+    assert m1.uid != m2.uid
+
+    # With a callable: we still should be able to set callables as defaults
+    class FunctionModel(BaseModel):
+        a: int = 1
+        uid: Callable[[], UUID] = Field(uuid4)
```
Comment (PrettyWood) [inline tests/test_main.py]: Had to use `= Field(uuid)` instead of `= uuid4` as functions are ignored when creating fields and the default is hence not properly set
I think this behaviour should be fixed but in another PR (maybe for v2 with the other syntax for dynamic values)

File: pydantic/fields.py [hunk 4]
```diff
@@ -97,8 +99,9 @@ def __init__(self, default: Any, **kwargs: Any) -> None:
 
 
 def Field(
-    default: Any,
+    default: Any = Undefined,
     *,
+    default_factory: Optional['Callable[[], Any]'] = None,
```
Comment (PrettyWood) [inline pydantic/fields.py]: Had to use `'Callable...'` because it would raise
`TypeError: 'ABCMeta' object is not subscriptable`

Comment (MrMrRobat) [reply pydantic/fields.py]: How is that possible? It already used in this module without an issues. That could be the case for collections.Callable, but it's a different thing.

Comment (PrettyWood) [reply pydantic/fields.py]: Ah right ! I didnt' see the `Callable` came from `.typing`. Thanks !

File: pydantic/fields.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 from typing import (
     TYPE_CHECKING,
     Any,
+    Callable,
```
Comment (samuelcolvin) [inline pydantic/fields.py]: if the below is `TypingCallable`, what is this? Maybe good to give this an alias.

File: pydantic/fields.py [hunk 6]
```diff
@@ -307,8 +321,9 @@ def prepare(self) -> None:
         Note: this method is **not** idempotent (because _type_analysis is not idempotent),
         e.g. calling it it multiple times may modify the field and configure it incorrectly.
         """
-        if self.default is not None and self.type_ is None:
-            self.type_ = type(self.default)
+        default_value = self.default_factory() if self.default_factory is not None else self.default
```
Comment (samuelcolvin) [inline pydantic/fields.py]: perhaps useful to have a `get_default()` function on this class which is used here and in `main.py` and perhaps deals with deepcopy?

File: pydantic/fields.py (hunk 5 above)
Comment (PrettyWood) [reply pydantic/fields.py]: I decided to add a `NoArgCallable` directly in the local `typing` module. Makes things clearer imo

**Comment by PrettyWood** [comment]
@samuelcolvin Thanks for the review ! Changes done

File: pydantic/fields.py [hunk 7]
```diff
@@ -272,13 +296,13 @@ def infer(
 
         if isinstance(value, FieldInfo):
             field_info = value
-            value = field_info.default
+            value = field_info.default_factory if field_info.default_factory is not None else field_info.default
```
Comment (samuelcolvin) [inline pydantic/fields.py]: This looks a bit weird here, shouldn't we use

```suggestion
            value = field_info.default_factory() if field_info.default_factory is not None else field_info.default
```

or otherwise modify the logic here? I know `value` being a callable won't break the logic below, but it's not what you would expect and could cause errors in future?

Comment (PrettyWood) [reply pydantic/fields.py]: Of course ! Sorry forgot to change it here

**Comment by jmagnusson** [comment]
@samuelcolvin When do you think a release will be ready with this PR? Looking forward to it!

**Comment by samuelcolvin** [comment]
@jmagnusson see #1341 for details on release date.
